cmd/kube-apiserver/app/testing/testserver.go

```diff
@@ -100,23 +102,15 @@ type TestServerInstanceOptions struct {
 
 // TestServer return values supplied by kube-test-ApiServer
 type TestServer struct {
-	ClientConfig      *restclient.Config        // Rest client config
+	ktesting.TContext
+	ClientConfig      *restclient.Config        // Rest client config, also available via TContext.RESTConfig() (= TestServer.RESTConfig())
```

Comment:
@mengjiao-liu, @aojea : style question...

My goal is to let `StartTestServer` handle the TContext and client set creation and tear-down. That makes it possible to remove some common boilerplate code that appears in all integration tests.

Should that TContext get returned inline as part of TestServer (as shown above) or as a separate return value?

When it's inline, test code looks like this:
```Go
func TestEndpointUpdates(t *testing.T) {
	// Disable ServiceAccount admission plugin as we don't have serviceaccount controller running.
	server := kubeapiservertesting.StartTestServerOrDie(t, nil, []string{"--disable-admission-plugins=ServiceAccount"}, framework.SharedEtcd())
	client := server.Client()
	informers := informers.NewSharedInformerFactory(client, 0)
	tCtx := server.TContext
        epController := endpoint.NewEndpointController(
                tCtx,
...
```

```patch
diff --git a/test/integration/endpoints/endpoints_test.go b/test/integration/endpoints/endpoints_test.go
index e1aee518c02..3cebed7b394 100644
--- a/test/integration/endpoints/endpoints_test.go
+++ b/test/integration/endpoints/endpoints_test.go
@@ -28,7 +28,6 @@ import (
        "k8s.io/apimachinery/pkg/util/intstr"
        "k8s.io/apimachinery/pkg/util/wait"
        "k8s.io/client-go/informers"
-       clientset "k8s.io/client-go/kubernetes"
        kubeapiservertesting "k8s.io/kubernetes/cmd/kube-apiserver/app/testing"
        "k8s.io/kubernetes/pkg/controller/endpoint"
        "k8s.io/kubernetes/test/integration/framework"
@@ -38,17 +37,10 @@ import (
 func TestEndpointUpdates(t *testing.T) {
        // Disable ServiceAccount admission plugin as we don't have serviceaccount controller running.
        server := kubeapiservertesting.StartTestServerOrDie(t, nil, []string{"--disable-admission-plugins=ServiceAccount"}, framework.SharedEtcd())
-       defer server.TearDownFn()
-
-       client, err := clientset.NewForConfig(server.ClientConfig)
-       if err != nil {
-               t.Fatalf("Error creating clientset: %v", err)
-       }
-
+       client := server.Client()
        informers := informers.NewSharedInformerFactory(client, 0)
 
-       tCtx := ktesting.Init(t)
-       defer tCtx.Cancel("test completed")
+       tCtx := server.TContext
        epController := endpoint.NewEndpointController(
                tCtx,
                informers.Core().V1().Pods(),
```


When it's separate:
```
func TestEndpointUpdates(t *testing.T) {
	// Disable ServiceAccount admission plugin as we don't have serviceaccount controller running.
	tCtx, server := kubeapiservertesting.StartTestServerOrDie(t, nil, []string{"--disable-admission-plugins=ServiceAccount"}, framework.SharedEtcd())
	client := tCtx.Client()
	informers := informers.NewSharedInformerFactory(client, 0)
        epController := endpoint.NewEndpointController(
                tCtx,
...
```

I'm leaning towards keeping it separate. The only downside is that all callers need to be updated, but I want to do that anyway.